File: backend/app/schemas/review.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import (
    FindingCategory,
    FindingDiffSide,
    FindingSeverity,
    FindingSource,
    FindingStatus,
    ReviewRisk,
    ReviewStatus,
    ReviewTriggerType,
)
from app.schemas.pull_request import PullRequestDetail


class ReviewCreate(BaseModel):
    pull_request_id: UUID
    commit_sha: str = Field(min_length=7, max_length=40)
    attempt_number: int = Field(default=1, ge=1)
    status: ReviewStatus = ReviewStatus.QUEUED
    overall_risk: ReviewRisk | None = None
    trigger_type: ReviewTriggerType
    model_name: str | None = Field(default=None, max_length=100)
    model_version: str | None = Field(default=None, max_length=100)
    prompt_version: str | None = Field(default=None, max_length=100)
    duration_ms: int | None = Field(default=None, ge=0)
    static_analysis_duration_ms: int | None = Field(default=None, ge=0)
    ai_analysis_duration_ms: int | None = Field(default=None, ge=0)
    input_tokens: int | None = Field(default=None, ge=0)
    output_tokens: int | None = Field(default=None, ge=0)
    total_tokens: int | None = Field(default=None, ge=0)
    estimated_cost_usd: Decimal | None = Field(default=None, ge=0)
    error_code: str | None = Field(default=None, max_length=100)
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class ReviewStatusUpdate(BaseModel):
    status: ReviewStatus
    overall_risk: ReviewRisk | None = None
    duration_ms: int | None = Field(default=None, ge=0)
    static_analysis_duration_ms: int | None = Field(default=None, ge=0)
    ai_analysis_duration_ms: int | None = Field(default=None, ge=0)
    input_tokens: int | None = Field(default=None, ge=0)
    output_tokens: int | None = Field(default=None, ge=0)
    total_tokens: int | None = Field(default=None, ge=0)
    estimated_cost_usd: Decimal | None = Field(default=None, ge=0)
    error_code: str | None = Field(default=None, max_length=100)
    error_message: str | None = None
    completed_at: datetime | None = None


class ReviewResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    pull_request_id: UUID
    commit_sha: str
    attempt_number: int
    status: ReviewStatus
    overall_risk: ReviewRisk | None
    trigger_type: ReviewTriggerType
    model_name: str | None
    model_version: str | None
    prompt_version: str | None
    duration_ms: int | None
    static_analysis_duration_ms: int | None
    ai_analysis_duration_ms: int | None
    input_tokens: int | None
    output_tokens: int | None
    total_tokens: int | None
    estimated_cost_usd: Decimal | None
    error_code: str | None
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None
    findings_count: int = 0
    high_severity_findings_count: int = 0
    created_at: datetime
    updated_at: datetime


class ReviewDetail(ReviewResponse):
    pull_request: PullRequestDetail


class ReviewFindingCreate(BaseModel):
    review_id: UUID
    file_path: str = Field(min_length=1, max_length=1024)
    start_line: int | None = Field(default=None, ge=1)
    end_line: int | None = Field(default=None, ge=1)
    diff_side: FindingDiffSide | None = None
    severity: FindingSeverity
    category: FindingCategory
    title: str = Field(min_length=1, max_length=255)
    problem: str = Field(min_length=1)
    explanation: str | None = None
    suggestion: str | None = None
    confidence: Decimal = Field(ge=0, le=1)
    source: FindingSource
    fingerprint: str = Field(min_length=64, max_length=64)
    code_snippet: str | None = None
    status: FindingStatus = FindingStatus.OPEN
    github_comment_id: int | None = Field(default=None, ge=1)
    published_to_github: bool = False

    @model_validator(mode="after")
    def validate_line_range(self) -> "ReviewFindingCreate":
        if self.start_line is not None and self.end_line is not None and self.end_line < self.start_line:
            raise ValueError("end_line cannot be less than start_line")
        return self


class ReviewFindingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    review_id: UUID
    file_path: str
    start_line: int | None
    end_line: int | None
    diff_side: FindingDiffSide | None
    severity: FindingSeverity
    category: FindingCategory
    title: str
    problem: str
    explanation: str | None
    suggestion: str | None
    confidence: Decimal
    source: FindingSource
    fingerprint: str
    code_snippet: str | None
    status: FindingStatus
    github_comment_id: int | None
    published_to_github: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/review.ts
import type { PullRequestDetail, PullRequestStatus } from "./pull-request.ts";

export const REVIEW_STATUSES = [
  "queued",
  "fetching",
  "static_analysis",
  "ai_analysis",
  "validating",
  "publishing",
  "completed",
  "failed",
] as const;
export type ReviewStatus = (typeof REVIEW_STATUSES)[number];

export const REVIEW_RISKS = ["none", "low", "medium", "high"] as const;
export type ReviewRisk = (typeof REVIEW_RISKS)[number];

export const REVIEW_TRIGGER_TYPES = [
  "manual",
  "webhook",
  "synchronization",
] as const;
export type ReviewTriggerType = (typeof REVIEW_TRIGGER_TYPES)[number];

export const FINDING_SEVERITIES = ["high", "medium", "low"] as const;
export type FindingSeverity = (typeof FINDING_SEVERITIES)[number];

export const FINDING_CATEGORIES = [
  "security",
  "bug",
  "validation",
  "error_handling",
  "performance",
  "database",
  "maintainability",
  "code_quality",
  "best_practice",
] as const;
export type FindingCategory = (typeof FINDING_CATEGORIES)[number];

export const FINDING_SOURCES = ["static", "ai", "hybrid"] as const;
export type FindingSource = (typeof FINDING_SOURCES)[number];

export const FINDING_STATUSES = ["open", "dismissed", "resolved"] as const;
export type FindingStatus = (typeof FINDING_STATUSES)[number];

export type FindingDiffSide = "left" | "right";

export interface Review {
  id: string;
  pull_request_id: string;
  repository_id: string | null;
  repository_full_name: string | null;
  pull_request_number: number | null;
  pull_request_title: string | null;
  pull_request_status: PullRequestStatus | null;
  commit_sha: string;
  attempt_number: number;
  status: ReviewStatus;
  overall_risk: ReviewRisk | null;
  trigger_type: ReviewTriggerType;
  model_name: string | null;
  model_version: string | null;
  prompt_version: string | null;
  duration_ms: number | null;
  static_analysis_duration_ms: number | null;
  ai_analysis_duration_ms: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  total_tokens: number | null;
  estimated_cost_usd: string | null;
  error_code: string | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  findings_count: number;
  high_severity_findings_count: number;
  created_at: string;
  updated_at: string;
}

export interface ReviewDetail extends Review {
  pull_request: PullRequestDetail;
}

export interface ReviewFinding {
  id: string;
  review_id: string;
  file_path: string;
  start_line: number | null;
  end_line: number | null;
  diff_side: FindingDiffSide | null;
  severity: FindingSeverity;
  category: FindingCategory;
  title: string;
  problem: string;
  explanation: string | null;
  suggestion: string | null;
  confidence: string;
  source: FindingSource;
  fingerprint: string;
  code_snippet: string | null;
  status: FindingStatus;
  github_comment_id: number | null;
  published_to_github: boolean;
  created_at: string;
  updated_at: string;
}
